connect/runtime/src/test/java/org/apache/kafka/connect/integration/BlockingConnectorTest.java

```diff
@@ -368,31 +374,54 @@ private static ConfigDef config() {
                 );
         }
 
+        /**
+         * {@link CountDownLatch#await() Wait} for the connector/task to reach the point in its lifecycle where
+         * it will block.
+         */
         public static void waitForBlock() throws InterruptedException, TimeoutException {
+            CountDownLatch awaitBlockLatch;
             synchronized (Block.class) {
-                if (blockLatch == null) {
-                    throw new IllegalArgumentException("No connector has been created yet");
-                }
+                awaitBlockLatch = Block.awaitBlockLatch;
+            }
+
+            if (awaitBlockLatch == null) {
+                throw new IllegalArgumentException("No connector has been created yet");
             }
 
             log.debug("Waiting for connector to block");
-            if (!blockLatch.await(CONNECTOR_BLOCK_TIMEOUT_MS, TimeUnit.MILLISECONDS)) {
+            if (!awaitBlockLatch.await(CONNECTOR_BLOCK_TIMEOUT_MS, TimeUnit.MILLISECONDS)) {
                 throw new TimeoutException("Timed out waiting for connector to block.");
```

Comment:
Since scanning creates connector instances, and validation caches the connector instance, how do you ensure that the right awaitBlockLatch is being waited on here?

Reply:
If you add this stanza before the `log.debug("Connector should now be blocked")` the tests still pass:

```
boolean retry;
synchronized (Block.class) {
    retry = Block.awaitBlockLatch != null && Block.awaitBlockLatch != awaitBlockLatch;
}
if (retry) {
    log.debug("New blocking instance was created, retrying wait");
    waitForBlock();
}
```
For me, I see this being printed in:
* testBlockInSinkTaskStart
* testBlockInConnectorStart
* testWorkerRestartWithBlockInConnectorStart
* testBlockInSourceTaskStart
* testBlockInConnectorInitialize

This leads me to believe that this function is normally exiting before the blocking method of the last-instantiated instance happens.

I don't immediately see how this could cause flakiness, but it's at least an instance of the method not doing what it says it does.

Reply:
WDYT about a `Block.prepare()` called before the test starts that creates the awaitBlockLatch, instead of having the Block constructor initialize it? That could eliminate the wait-notify mechanism, since only one thread (the test thread) would be responsible for setting/clearing the awaitBlockLatch.

Reply:
I don't believe these changes are necessary here since the portions they address are not affected by the PR. If you would like to do this cleanup in a separate PR, I'd be happy to review.